x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java

```diff
@@ -176,19 +177,35 @@ public static TypeResolution isType(
         ParamOrdinal paramOrd,
         String... acceptedTypes
     ) {
-        return predicate.test(e.dataType()) || e.dataType() == NULL
-            ? TypeResolution.TYPE_RESOLVED
-            : new TypeResolution(
-                format(
-                    null,
-                    "{}argument of [{}] must be [{}], found value [{}] type [{}]",
-                    paramOrd == null || paramOrd == DEFAULT ? "" : paramOrd.name().toLowerCase(Locale.ROOT) + " ",
-                    operationName,
-                    acceptedTypesForErrorMsg(acceptedTypes),
-                    name(e),
-                    e.dataType().typeName()
-                )
-            );
+        if (predicate.test(e.dataType()) || e.dataType() == NULL) {
+            return TypeResolution.TYPE_RESOLVED;
+        }
+
+        if (e instanceof FieldAttribute fa && fa.field() instanceof InvalidMappedField imf) {
+            return new TypeResolution(errorStringIncompatibleTypes(operationName, paramOrd, name(e), imf.errorMessage(), acceptedTypes));
+        }
```

Comment:
This improves error messages for failed type resolutions in case of union types, see yaml tests.